Fix CIDR to World Entity Conversion Bug

nathanjsweet opened the pull request:
Previously if either the "0.0.0.0/0" CIDR (hereafter "world-ipv4")
or a world-ipv6 (i.e. "::/0") CIDR was present the selector creation logic
for CIDRs would also add the "reserved:world" entity to the
selector labels generated by the CIDR for the sake of the policy engine.
    
This does not work when Cilium is dual-stack mode as the "reserved:world"
entity covers both IPv4 and IPv6, so both CIDRs should be present before such
an addition to the policy engine logic is made. Moreover, dual-stack mode lacks
any mechanism to distinguish world-ipv4 and world-ipv6 from one another. 
This bug has prevented the possibility of managing world-ipv4 and world-ipv6
separately from one another in network policies.
    
Separate identities (and labels, etc) for world-ipv4 and world-ipv6 now exist
so they can be distinguished.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

Fixes: #22610 
```release-note
bug: In dual-stack mode (both IPv4 and IPv6 are enabled), Cilium incorrectly converted CIDRs that covered all possible addresses for an IP Family (e.g. 0.0.0.0/0) to the "reserved:world" entity. Both IP families must be completely covered for "reserved:world" to apply. This resulted in dual-stack mode network policies that could not distinguish between world IPv4 and IPv6 traffic, treating them as one entity instead.
```

aanm reviewed (approved):
LGTM @nathanjsweet just a heads-up that since we have branched v1.13, this PR will not be part of the 1.13 unless it is set with the needs-backport/1.13 label.

joestringer reviewed (approved):
Thanks for the fix! LGTM.

pchaigno commented on pkg/policy/api/cidr.go:
```diff
@@ -75,19 +67,24 @@ func (s CIDRSlice) GetAsEndpointSelectors() EndpointSelectorSlice {
 		}
 		// TODO: Log the error?
 	}
+	if hasIPv4AllBeenAdded && hasIPv6AllBeenAdded {
```
Couldn't we take into account whether we are running an IPv4-only or IPv6-only cluster? As Joe mentioned offline, this fix may have a big impact on the policy map size because that one entry will now be converted into one entry per existing CIDR IIUC. So if we can limit its effect to dual-stack clusters, that would probably be best.

joestringer commented on pkg/policy/api/cidr.go:
```diff
@@ -75,19 +68,31 @@ func (s CIDRSlice) GetAsEndpointSelectors() EndpointSelectorSlice {
 		// TODO: Log the error?
 	}
+	// We want to add the world selector whenever possible as it limits the growth of the
+	// ipcache.
```
Now I'm second-guessing whether there's really an optimization to make here or not. The thinking I had when I raised the question was whether the "allow to world" could become an "allow all" match, and hence be encoded with one entry in the datapath. But in reality we can only inject that match if the policy says "toEntities: all", ie matching also identities within the cluster. Maybe the best that this will achieve is just to shortcut some of the selection logic for the case where it's single stack and the broadest CIDR is allowed by a policy.

I'm not sure this will actually impact either ipcache or policymaps size at this point.

joestringer replied:
(The logic is probably fine either way, the latest version here is a little more complicated but still correct. It's just the comment that I think may not be accurate).

pchaigno replied:
Ah, yes. That's not going to impact the policy map size because we *append* the world entity to the CIDR slice.

But then, why do we even do that in the first place? Doesn't the 0.0.0.0/0 CIDR identity already match all traffic?

(I'm a bit worried that, if we don't do that to optimize, then maybe it's nothing we can't just skip doing. And maybe it explains the previous CI failure in ConformanceDatapath.)

joestringer replied:
yeah I'd expect `0.0.0.0/0` CIDR identity to match all (IPv4) traffic already. Maybe we didn't need the extra world selector insertion after all, we just never revisited this. I think it's important that the _identities_ we create have the `reserved:world` label to make sure that any `toEntities: world` policy matches those identities, but I don't think we need to inject the `reserved:world` labelSelector into the _policy parsing_ piece for the selectors that then select the identities.

pchaigno reviewed (dismissed):
Blocking on above conversation with Joe and CI failure. I suspect the two are related :slightly_smiling_face:

joestringer reviewed (commented):
Could you add an upgrade guide note that a CIDR match for `0.0.0.0/0` no longer allows the CIDR `::/0` for IPv6 traffic?

nathanjsweet committed:
Fix CIDR to World Entity Conversion Bug

Previously if either the "0.0.0.0/0" CIDR (hereafter "world-ipv4")
or a world-ipv6 CIDR was present the selector creation logic
for CIDRs would also add the "reserved:world" entity to the
selector labels generated by the CIDR.

This does not work when Cilium is dual-stack mode as "reserved:world"
covers both IPv4 and IPv6, so both CIDRs should be present before such
an addition is made. Moreover, dual-stack mode lacked any mechanism to
distinguish world-ipv4 and world-ipv6 from another. This bug has prevented
the possibility of managing world-ipv4 and world-ipv6 separately from one
another in the policy engine.

This inccorect assumption goes all the way to the datapath, where SECLABEL
and WORLD_ID were part of drop notification, policy logic, and tunnel logic.
SECLABLE and WORLD_ID were both set by userspace compilation logic.

Separate identities (and labels, etc) for world-ipv4 and world-ipv6 now exist
so they can be distinguished in userspace and in the datapath.

Some utility methods have been added in userspace and datapath to check
for the necessity of using world-ipv4 and world-ipv6 identities.

Explicit tests for dual-stack mode have been added.

The generic WORLD_ID is still used as the VNI tag for all tunnel logic
in the datapath. When tunnel code is encapsulating any WORLD_IPX_ID
identity it will convert the VNI to WORLD_ID. When tunnel code is
decapsulating any WORLD_ID VNI it will convert the VNI to either
WORLD_IPV4_ID or WORLD_IPV6_ID based on the underlying packet
protocol.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>

chancez commented on pkg/hubble/metrics/flows-to-world/handler.go:
```diff
@@ -53,7 +53,10 @@ func (h *flowsToWorldHandler) Init(registry *prometheus.Registry, options api.Op
 		Help:      "Total number of flows to reserved:world",
 	}, labels)
-	h.worldLabel = fmt.Sprintf("%s:%s", pkglabels.LabelSourceReserved, pkglabels.IDNameWorld)
+	h.worldLabels = make(map[string]struct{})
```
Any reason to use a map over something simpler like a switch statement? Also, perhaps we should move `isReservedWorld` into `pkg/identity/numericidentity.go`? It doesn't seem like it needs to be a method, and checking if something is "world" seems like something we should be able to reuse.

nathanjsweet replied:
I didn't want to modify this logic too much. I don't know why this `worldLabels` solution was chosen, perhaps for no particular reason.  If there is another compelling reason to modify this PR then I will try to refactor this logic.

chancez replied:
Gotcha. Then maybe we can just move these labels to 3 global `var` and use a switch/if instead of a map? I think that aligns more closely with the previous logic, and avoids the map allocation.

aanm commented on pkg/labels/labels.go:
```diff
@@ -542,10 +556,12 @@ func (l Labels) IsReserved() bool {
+		for _, lbl := range l {
+			if lbl.matches(&label) {
+				return true
+			}
```
I don't understand this change and why it's necessary. Can you clarify it? Thank you

nathanjsweet replied:
It gets used [here](https://github.com/cilium/cilium/pull/22625/files#diff-bb1165de1a65911d59ac7f9a002aa9f1e25c30e073547adec11b174a174fdca6R431).

nathanjsweet replied:
I suppose I could just or together the `Has` invocations.

chancez reviewed (approved):
LGTM

joestringer reviewed (approved):
This came out really tidy, nice work.

gandro reviewed (approved):
Thanks a lot for tackling this! My code owners look good to me. We should probably document this in the upgrade notes too.

christarazi reviewed (changes requested):
I think we need to block on my previous [comment](https://github.com/cilium/cilium/pull/22625#issuecomment-1560234953) until we have datapath experts chime in and says it's fine.

christarazi commented on bpf/node_config.h:
```diff
@@ -47,6 +47,13 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 
 #define HOST_ID 1
 #define WORLD_ID 2
+#if defined ENABLE_IPV4 && defined ENABLE_IPV6
+# define WORLD_IPV4_ID 9
+# define WORLD_IPV6_ID 10
+#else
+# define WORLD_IPV4_ID 2
+# define WORLD_IPV6_ID 2
+#endif
```
Nice :+1:

lmb commented on bpf/lib/nodeport.h:
```diff
@@ -304,7 +304,7 @@ static __always_inline int encap_geneve_dsr_opt6(struct __ctx_buff *ctx,
 						src_port,
 						tunnel_endpoint,
-						WORLD_ID,
+						WORLD_IPV4_ID,
```
To be on the safe side: the function seems to be dealing with v6, but here we want the v4 id?

nathanjsweet replied:
Good catch. I double-checked all the other references to `WORLD_*` in this PR. They are all correct. I think this was a rebase error (the PR modifies high-commit files, so there have been numerous rebases).

pchaigno reviewed (approved):
IIUC, when in dual-stack mode and fully upgraded, the datapath should not have any reference to `WORLD_ID` anymore. Does that mean we can remove the `WORLD_ID` entry from the policy map for entity world in v1.15?

In dual-stack mode, there is one reference to `WORLD_ID` left at the beginning of `handle_netdev`, in the `send_drop_notify` call.

Regarding the upgrade impact, I've seen https://github.com/cilium/cilium/pull/25204/files. For dual-stack clusters using the `world` entity in policies, I would expect packet drops or policy bypasses if the ipcache catch-all entry is updated (from 0.0.0.0/0=>2 to 0.0.0.0/0=>9) before the bpf_lxc programs and their policy maps.

How was the upgrade tested? Were such policies loaded? @joestringer Do you know how the ipcache is updated in such situations?

Regarding downgrade, couldn't we avoid the connectivity breakage with a patch in v1.13?

Policy changes with upgrade impacts are very tricky. Nice work driving this through!

asauber commented on bpf/TAGS:
was this tags file committed intentionally?

nathanjsweet replied:
Yeah.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -583,6 +583,19 @@ func (l4Filter *L4Filter) toMapState(p *EndpointPolicy, identities Identities, f
 			keyToAdd.Identity = id.Uint32()
 			if entryCb(keyToAdd, &entry) {
 				p.PolicyMapState.denyPreferredInsertWithChanges(keyToAdd, entry, identities, features, changes)
+				// If Cilium is in dual-stack mode then the "World" identity
+				// needs to be split into two identities to represent World
+				// IPv6 and IPv4 traffic distinctly from one another.
+				if id == identity.ReservedIdentityWorld && option.Config.IsDualStack() {
```
Why doesn't `cs.GetSelections()` already select both of the world identities if world was selected?

joestringer replied:
^ Answer is that the `toEntities: - world` selector doesn't select the world-ipv4 / world-ipv6 identities currently, see my previous comments above. I'd argue that we should just put those labels into the world selector, then this code doesn't need to be handled as a special case here in `toMapState()`.

nathanjsweet replied:
I prefer to do that in another PR. Making these labels selectable **could** open up a can of worms.

joestringer replied:
Please do follow up. What concerns me is that if we have to place logic like this here, then there could be other places in the policy engine where we're handling world identity and this logic is missing. Embedding the selection into the selectors seems like a more comprehensive way to solve this issue without having to add this extra logic here.

nathanjsweet replied:
Done: https://github.com/cilium/cilium/issues/27494

squeed replied:
We did miss this logic; world is a label, not an identity. https://github.com/cilium/cilium/pull/29958 fixes this.

We might be able to remove this code now.

joestringer commented on bpf/node_config.h:
```diff
@@ -49,6 +49,13 @@ DEFINE_IPV6(HOST_IP, 0xbe, 0xef, 0x0, 0x0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x0, 0xa, 0x
 
 #define HOST_ID 1
 #define WORLD_ID 2
```
Should we drop WORLD_ID now to ensure that all code respects the protocol-specific identities?

joestringer replied:
Ah I guess the idea is that even in dual-stack the WORLD_ID handles traffic that isn't IPv4 or IPv6.

My general concern here is that new code might naively continue to use `WORLD_ID` when it should actually use `WORLD_ID_IPV4`/v6 and this could introduce weird bugs. Would it make sense to rename `WORLD_ID` to `WORLD_ID_NO_PROTO` with a comment describing to datapath folks how it should only be used for traffic handling when the traffic is not IPv4/IPv6?

nathanjsweet replied:
We need to have a broader conversation about this with datapath folks. I would categorize this as a possible next step, but it changes the UX of the policy engine. We should do that in another PR.

joestringer replied:
Why does it change the UX of policy engine? It seems like this PR already splits world into three identities depending on L3 protocol, so if this has any repercussions then I think that's important to raise & engage with, since folks will need to develop future datapath changes based on that understanding.

(I didn't see anything particularly concerning in this PR and I don't think that this needs to block merging this bugfix, but I do want to make sure this discussion happens with datapath folks).

nathanjsweet replied:
> Why does it change the UX of policy engine?

Well, maybe not directly, but these entities could be selectable. Of course we could figure out a way to not expose them as referenceable entities in a policy file.

joestringer replied:
I don't think they need to be user-facing entities. It's sufficient for these _identities_ selected by the existing policy APIs, as an internal detail.

joestringer replied:
^ No reason to block this PR on this discussion point, but we can continue to discuss out-of-band on the implications here and longer term supportability.

joestringer commented on pkg/datapath/linux/config/config.go:
```diff
@@ -1014,6 +1022,8 @@ func (h *HeaderfileWriter) writeStaticData(fw io.Writer, e datapath.EndpointConf
 
 	secID := e.GetIdentityLocked().Uint32()
 	fmt.Fprint(fw, defineUint32("SECLABEL", secID))
+	fmt.Fprint(fw, defineUint32("SECLABEL_IPV4", secID))
+	fmt.Fprint(fw, defineUint32("SECLABEL_IPV6", secID))
```
Why is this necessary?

I was under the impression that SECLABEL is always the endpoint's security identity, which should not change depending on the protocol. Given that this code is serializing it exactly the same as `SECLABEL`, what's the difference to the datapath handling?

nathanjsweet replied:
`SECLABEL_IPV4` and `SECLABEL_IPV6` are used in most cases in the `/bpf` code now so they have to be defined, even when they are not "relevant". Cases where they are relevant are the "bpf_host.c", "bpf_lxc.c", and "bpf_xdp.c" programs, where `SECLABEL` is compiled into the program as `WORLD_ID` (see `loader/base.go` as an example). In dual-stack mode this needs to be on a per IP-version basis.

joestringer replied:
I see. I'm not sure I logically follow how SECLABEL is now used in those cases, but I buy your argument :+1: . I missed the loader/base.go handling.

joestringer commented on pkg/policy/api/selector.go:
```diff
@@ -234,6 +234,8 @@ var (
 		labels.IDNameHost:       newReservedEndpointSelector(labels.IDNameHost),
 		labels.IDNameRemoteNode: newReservedEndpointSelector(labels.IDNameRemoteNode),
 		labels.IDNameWorld:      newReservedEndpointSelector(labels.IDNameWorld),
+		labels.IDNameWorldIPv4:  newReservedEndpointSelector(labels.IDNameWorldIPv4),
+		labels.IDNameWorldIPv6:  newReservedEndpointSelector(labels.IDNameWorldIPv6),
```
Shouldn't the world reserved selector also select IPv4 and IPv6 world labels now?

The issue IIRC is that a `0.0.0.0/0` CIDR policy was implicitly translated into a world selector, but it should really be translated to world-ipv4 selector. However, if a user actually adds `ToEntities: - world` then that should select both IPv4 and IPv6 destinations, no?

joestringer replied:
I guess the answer to this question depends on whether the new reserved identities are declared with the world label in addition to the world-$proto or not.

joestringer replied:
Ah I see in `pkg/identity/numericidentity.go` the proto-specific world identities don't have the `reserved:world` label in them.

joestringer replied:
(If we do change this, it looks like it needs to happen here and in `entity.go`.)

nathanjsweet replied:
Making those selectable is out-of-scope for this PR. That's something we can certainly explore in the future.

joestringer replied:
(same thread as https://github.com/cilium/cilium/pull/22625#discussion_r1293797898, resolving per that comment)

joestringer commented on bpf/lib/identity.h:
```diff
@@ -137,7 +185,18 @@ static __always_inline __u32 inherit_identity_from_host(struct __ctx_buff *ctx,
 #endif
 	} else {
+#if defined ENABLE_IPV4 && defined ENABLE_IPV6
+		__u16 proto = ctx_get_protocol(ctx);
+
+		if (proto == bpf_htons(ETH_P_IP))
+			*identity = WORLD_IPV4_ID;
+		else if (proto == bpf_htons(ETH_P_IPV6))
+			*identity = WORLD_IPV6_ID;
+		else
+			*identity = WORLD_ID;
```
nit: We could probably factor this into a function to share between here and `bpf_host.c`, like `world_identity(proto)`.

joestringer replied:
^ We can always follow up on this too, it's a nit.

joestringer reviewed (commented):
I re-reviewed for sig-agent and while I was at it, looked back over the whole PR. I'm not sure I understand why `SECLABEL` needs to be split by protocol, we can discuss in the thread.

Other than that there's just a couple of potential minor cleanups like if we make the world selector select the new world-ipv4 / world-ipv6 identities then we can remove a special case in `toMapState()`. I also saw a small piece of reusable code in the datapath that could potentially be defined once and reused.

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -487,8 +487,8 @@ int tail_handle_arp(struct __ctx_buff *ctx)
-						       LOCAL_NODE_ID, WORLD_ID,
-						       WORLD_ID, &trace);
+						       LOCAL_NODE_ID, WORLD_IPV4_ID,
+						       WORLD_IPV4_ID, &trace);
```
Hmm, setting the VNI field in the packet to `WORLD_IPV4_ID` could potentially cause confusion for other agents in the cluster when transmitting these packets, until those peer agents have also upgraded to a version that understands `WORLD_IPV4_ID`. How is this code exercised?

joestringer replied:
Probably what this should do is transmit WORLD_ID and then the receive side should decide to treat "WORLD_ID" as "WORLD_IPV4_ID" iff dualstack is enabled at the destination. That would ensure that this doesn't cause connectivity issues if the peer node doesn't understand WORLD_IPV4_ID upon upgrade.

joestringer replied:
(discovered while freshly looking at the multiprotocol SECLABEL handling in the PR, but I didn't search around for whether there might be any other similar cases)

nathanjsweet replied:
So it will only be for dual-stack. It is an upgrade issue that I hadn't considered, but I don't see any way around it.

joestringer replied:
Hmm, digging back through the git history, it looks like this encap identity here is specifically for the VTEP support. Using any VNI other than 2 may just break that feature. I think the right solution here is to keep the second to last parameter as `WORLD_ID`. The intermediate parameter can stay as `WORLD_IPV4_ID`, since that is used for local trace notifications only.

Even if we wanted to transition this to transmitting a different identity, I think that in general we should be introducing that gradually to minimize cluster disruption during the upgrades. But in this particular case with VTEP I don't even see a path for that.

nathanjsweet replied:
[It looks like the `tunnel_id` is also set to the SECLABEL when VTEP is not enabled](https://github.com/cilium/cilium/blob/b3a19b54269435248419d2cbe6ff42afd9b479b7/bpf/lib/overloadable_skb.h#L202). Woulnd't that also be a problem?

joestringer replied:
Hmm yeah I didn't step deep enough. 

My gut feeling here is that the right approach for the backportable fix is to set both to WORLD_ID. Even if we wanted to start transmitting the newer identity, that change should only be introduced in v1.15 after we backported this [PR transmitting `WORLD_ID`] to at least v1.14.

Then it's a question of exactly which paths care about handling ARP from a tunnel peer. I think the reason it's important in a VTEP case is that the VTEP doesn't have another out-of-bound mechanism to learn Pod IP locations. However I'm not sure whether Cilium ever sends ARP like this over the tunnels. If this is purely for VTEP then in this case I think it's both (a) fine to leave these as WORLD_ID and (b) there's no further changes necessary, it's the correct fix.

On the other hand if there is a non-VTEP case where this path is expected to do ARP response then the peer's handling of the security identity should be further inspected.

nathanjsweet replied:
I'll try this out, but I got tunneling errors in CI before I made the SECLABEL changes.

joestringer replied:
Those tunneling errors could potentially be the first sign of upgrade compatibility impact, because new identities were being transmitted and the corresponding receive paths were not handling them. Hard to say without seeing them directly though.

joestringer replied:
The approach in the latest changes LGTM in principle for solving the issue raised by this thread, there's just another couple of missing cases. I'll mark this thread as resolved.

christarazi commented on pkg/fqdn/dnsproxy/proxy.go:
```diff
@@ -851,6 +851,13 @@ func (p *DNSProxy) ServeDNS(w dns.ResponseWriter, request *dns.Msg) {
 
 	targetServerID := identity.ReservedIdentityWorld
+	if option.Config.IsDualStack() {
+		if targetServerIP.To4() == nil {
```
If `To4()` returns nil, then we set the ID to world-ipv4? Shouldn't the condition be

```suggestion
		if targetServerIP.To4() != nil {
```

[1]

nathanjsweet replied:
Good catch.

christarazi commented on pkg/proxy/logger/endpoint/epinfo.go:
```diff
@@ -73,7 +74,15 @@ func (r *endpointInfoRegistry) FillEndpointInfo(info *accesslog.EndpointInfo, ip
+				}
+			} else {
+				id = identity.ReservedIdentityWorld
+			}
```
Maybe it's best to define a util function somewhere that returns the identity based on the IP, so that we keep the code DRY and prevent bugs such as [1]

christarazi reviewed (changes requested):
Reviewed just proxy-related code files.

christarazi reviewed (approved):
LGTM from proxy side

joestringer reviewed (changes requested):
The latest changes for encap look good, and there's one decap case handled, then I think there's a couple more remaining decap cases in `bpf/bpf_overlay.c:handle_ipv4()` and `bpf/bpf_overlay.c:handle_ipv6()` , both for the high scale and regular tunnel decap cases without encryption.

These suggest to me that we don't have policy coverage for an IPv4-only `0.0.0.0/0` allow for pods where the traffic is received from outside the cluster through a nodeport on a different node. That sort of test should be able to highlight a policy deny issue with this latest implementation.

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -75,13 +75,15 @@ static __always_inline int handle_ipv6(struct __ctx_buff *ctx,
-		if (info)
-			*identity = key.tunnel_id = info->sec_identity;
+		if (info) {
+			*identity = info->sec_identity;
+			key.tunnel_id = get_tunnel_id(info->sec_identity);
```
I don't think that we need to change this case since the security identity is coming from ipcache?

Even if we need to change this line, I don't think that `*identity` and `key.tunnel_id` should diverge.

nathanjsweet replied:
An ipcache lookup could still result in the identity being `WORLD_IPV6_ID`.

joestringer replied:
That's correct on receive, right? Internally the datapath would understand `WORLD_IPV6_ID`, it's only when we serialize/deserialize we need to understand that peers may be transmitting `WORLD_ID`.

joestringer replied:
^ Reviewing afresh with Nathan,this path is updating `key.tunnel_id` to ensure that whenever we modify that field, we always encode `WORLD_IPV*_ID` into `WORLD_ID`, and whenever decoding from that field, we convert back.

I'm still a little bit concerned about the potential for divergence here, but this is probably easier to ensure that the implementation is correct because any subsequent reads from the `tunnel_id` field will convert back. I withdraw my comment here :+1:

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -321,16 +323,18 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
-		if (info)
-			*identity = key.tunnel_id = info->sec_identity;
+		if (info) {
+			*identity = info->sec_identity;
+			key.tunnel_id = get_tunnel_id(info->sec_identity);
```
Same comment as IPv6 path.

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -321,16 +323,18 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 	} else {
 #ifdef ENABLE_HIGH_SCALE_IPCACHE
-		key.tunnel_id = *identity;
+		key.tunnel_id = get_tunnel_id(*identity);
```
This should also be `get_id_from_tunnel_id()` right? Effectively that function is the "receive" side converter, while `get_tunnel_id()` is the transmit side?
```suggestion
		key.tunnel_id = get_id_from_tunnel_id(*identity);
```

nathanjsweet replied:
No. Any tunnel key look-up should always convert `WORLD_IPVX_ID` to `WORLD_ID`. It doesn't matter if its on the receive-side.

joestringer replied:
I disagree. What's important is that on the wire, `WORLD_ID` is used to ensure compatibility with older releases. At the same time, internally in the datapath if running in dual-stack mode, the datapath should consistently identify the identity of that peer and if it's an IPv4-based world identity peer that should be `WORLD_IPV4_ID`. That way, when we receive traffic with `WORLD_ID`, we can convert that to `WORLD_IPV4_ID`, then generate hubble flows with `WORLD_IPV4_ID` and do network policy lookups with `WORLD_IPV4_ID` in order to properly reflect that this new version of Cilium is identifying this traffic as a more narrowly scoped "world" for the specific protocol.

At the minimum, the policy lookup needs to use the protocol-scoped world identity to perform the lookup, otherwise the traffic could be denied by an "allow `0.0.0.0/0`" policy (which is encoded in the policymap as "allow from identity `9`" in dual-stack scenarios). If we're doing that for policy lookup, then the datapath should likewise emit all hubble events with the identity that the datapath is considering this traffic to correspond to. Otherwise we'll have a lot of difficulty debugging problems related to these new protocol-specific world identities.

joestringer replied:
Reading through this afresh, I assumed that with the `#ifdef ENABLE_HIGH_SCALE_IPCACHE` function that the macro if/else was updating the same variables / fields and that this line was updating `*identity`. I then argued above that internally within the datapath we should always be using the newer protocol-specific world identities. The first assumption was wrong and this code is enforcing the exact property that I was arguing for. :sweat_smile: 

Sorry for the confusion! LGTM.

joestringer replied:
(I'm now a bit confused why the high scale ipcache code here doesn't update `*identity`, but that has nothing to do with this PR).

joestringer commented on bpf/lib/overloadable_skb.h:
```diff
@@ -183,6 +183,30 @@ static __always_inline bool ctx_snat_done(const struct __sk_buff *ctx)
 	return (ctx->mark & MARK_MAGIC_HOST_MASK) == MARK_MAGIC_SNAT_DONE;
 }
 
+static __always_inline __u32 get_tunnel_id(__u32 identity)
```
Why do we need to define these functions twice? There is nothing specific to skb or xdp buffers here. We should be able to define these functions once in `lib/identity.h`?

nathanjsweet replied:
It can't live in `identity.h` as `identity.h` implicitly requires `overloadable_xxx.h`.

joestringer replied:
Ah right. I guess `lib/common.h` is an option to avoid duplication.

joestringer commented on bpf/bpf_overlay.c:
```diff
@@ -345,7 +349,7 @@ static __always_inline int handle_ipv4(struct __ctx_buff *ctx,
 			if (!vtep)
 				goto skip_vtep;
 			if (vtep->tunnel_endpoint) {
-				if (*identity != WORLD_ID)
+				if (identity_is_world_ipv4(*identity))
```
This looks like it reversed the logic, breaking VTEP support, see https://github.com/cilium/cilium/issues/31023.
